### PR TITLE
new(libsinsp): Expand concept of process group leader attributes for filtering (new `proc.vpgid.name`, `proc.vpgid.exe`, `proc.vpgid.exepath`, `proc.is_sid_leader`, `proc.is_vpgid_leader`)

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Similar to how we support the concept of session leader expand support for concept of process group leader.

Can help to distinguish if the process' cmdline was 'directly' executed for instance in a tty (similar to bash history logging, `is_vpgid_leader` would be 'true') or executed as descendent process in the same process group.

One step closer to the concept of "manual command" logging when in an SSH session or Terminal shell session. The limitation of shell built-ins not being logged, as no new process is created in such cases, still applies, of course.

It could be utilized for various threat detection purposes or as a general logging filter to better distinguish, for example, what was spawned as a result of executing a script, such as some bash pipes within a script (e.g., `bash -c 'echo 123456789 | base64'`), or even for detecting command injection. In such cases `is_vpgid_leader` would be 'false'.

Still thinking if and how useful it could be to build stronger Reverse Shell detections in scenarios where certain gaps exist.

Lastly, aligns with ongoing efforts to make the process tree, or more generally, the process hierarchy, more robust.

https://www.win.tue.nl/~aeb/linux/lk/lk-10.html

**Which issue(s) this PR fixes**:

Not fixing, but related to https://github.com/falcosecurity/falco/issues/2338 and other issues.
Especially from new adopters, I hear the desire to log all interactive commands as the number one use case to get started.

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(libsinsp): Expand concept of process group leader attributes for filtering (new `proc.vpgid.name`, `proc.vpgid.exe`, `proc.vpgid.exepath`, `proc.is_sid_leader`, `proc.is_vpgid_leader`)
```

### CODE REVIEW

**Comment by incertum** [comment]
Would rate our upstream rules capabilities, in terms of handling a wide range of RCE and command injection variations, as not yet sufficient. Started experimenting to explore if and how this could assist in creating more effective rules, as we enhance the semantic understanding of individual processes to some extent (beyond just the parent process lineage information). If not helpful for RCE rules it still helps for use cases mentioned above.

@loresuso @LucaGuerra @lrishi @allanembedded @darryk10  @leogr perhaps you would have some thoughts or ideas?
Also CC @Andreagit97 as its related to the cleanup and extension of process hierarchies.

File: userspace/libsinsp/filterchecks.cpp [hunk 1]
```diff
@@ -2782,6 +2787,138 @@ uint8_t* sinsp_filter_check_thread::extract(sinsp_evt *evt, OUT uint32_t* len, b
 				RETURN_EXTRACT_STRING(m_tstr);
 			}
 		}
+		case TYPE_VPGID_NAME:
+		{
+			//
+			// Relying on the convention that a process group id is the process id of the process group leader
+			//
+			sinsp_threadinfo* vpgidinfo =
```
Comment (Andreagit97) [inline userspace/libsinsp/filterchecks.cpp:2795]: uhm maybe I'm missing something but here we are accessing a table indexed by tid referred to the init namespace, with a tid referred to a possible different namespace :thinking:

File: userspace/libsinsp/test/events_proc.ut.cpp [hunk 2]
```diff
@@ -355,6 +356,14 @@ TEST_F(sinsp_with_test_input, spawn_process)
 	ASSERT_EQ(get_field_as_string(evt, "proc.sname"), "init");
 	ASSERT_EQ(get_field_as_string(evt, "proc.sid.exe"), "/sbin/init");
 	ASSERT_EQ(get_field_as_string(evt, "proc.sid.exepath"), "/sbin/init");
+	ASSERT_EQ(get_field_as_string(evt, "proc.is_sid_leader"), "false");
```
Comment (Andreagit97) [inline userspace/libsinsp/test/events_proc.ut.cpp:359]: Of course, I appreciate all the tests but I think we need to go a step further in what concern the testing. If we put ourselves in the simple case we won't discover possible failures, we probably need a real tree on which we can assert these values. Here i proposed a first attempt https://github.com/falcosecurity/libs/pull/1182/files#diff-505fbdbaff39fa9ce959ca0c186a07c3aaa89c4c3850b5d7913e74edcfab923a of course the tree is still simple but at least we have a namespace and we can try to assert something

Comment (Andreagit97) [reply userspace/libsinsp/test/events_proc.ut.cpp:359]: this is not related to this PR in particular but it is more of a generic comment, we probably need a generic process tree with 2/3 namespaces on which we assert all these process id values

Comment (Andreagit97) [reply userspace/libsinsp/test/events_proc.ut.cpp:359]: I would say after the other PR introduce them :+1: but yes we can start thinking about how to expand them. 
Right now they are mainly focused on 3 fields: tid, pid, ptid, but they should be extensible to other fields like vtid, vpid, vpgid, ecc... 
I have 2 main ideas in mind:
1.  Enrich the DEFAULT TREE with other processes and maybe another namespace and allow assertion on many different fields
2. Produce scap-fileS (:rofl: ) and parse them matching the actual system with a pstree snapshot as I've done here https://github.com/falcosecurity/libs/pull/1182/files#diff-e62ccb1906fa6f9ad858340e2d9b177b413a55e4a2dd573527265d3c8b84e812

(1) should help us in asserting that if the driver sends correct data userspace behaves well
(2) should help in understanding if our sinsp state reflects the running system

File: userspace/libsinsp/filterchecks.cpp (hunk 1 above)
Comment (incertum) [reply userspace/libsinsp/filterchecks.cpp]: @Andreagit97 I investigated why it was working and you will see it in the last commit. On top I added a clearer distinction between host processes and pid namespace processes. Proposing to also address this for `m_sid` as maybe today we could potentially fetch the wrong information in some edge cases? Let me know and I can add that commit. Thanks.

File: userspace/libsinsp/test/events_proc.ut.cpp (hunk 2 above)
Comment (incertum) [reply userspace/libsinsp/test/events_proc.ut.cpp:359]: I am curious now :smile: Meanwhile here some good old JSON lines of me running a py script in a container.

```
What I typed and executed from my terminal in a container

root@05df91a3e423:/test# whoami
root@05df91a3e423:/test# python3 memfd.p

{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"whoami","proc.exepath":"/usr/bin/whoami","proc.is_sid_leader":false,"proc.is_vpgid_leader":true,"proc.pid":947514,"proc.ppid":901354,"proc.pvpid":1,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":847,"proc.vpgid.exe":"whoami","proc.vpgid.exepath":"/usr/bin/whoami","proc.vpgid.name":"whoami","proc.vpid":847}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"python3 memfd.py","proc.exepath":"/usr/bin/python3","proc.is_sid_leader":false,"proc.is_vpgid_leader":true,"proc.pid":947555,"proc.ppid":901354,"proc.pvpid":1,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":848}

{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"ldconfig /sbin/ldconfig -p","proc.exepath":"/sbin/ldconfig","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947556,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":849}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"ldconfig.real -p","proc.exepath":"/sbin/ldconfig.real","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947556,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":849}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"sh -c bash -c 'exec ls -l /proc/848/fd/'","proc.exepath":"/bin/sh","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947557,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":850}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"bash -c exec ls -l /proc/848/fd/","proc.exepath":"/usr/bin/bash","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947558,"proc.ppid":947557,"proc.pvpid":850,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":851}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"ls -l /proc/848/fd/","proc.exepath":"/usr/bin/ls","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947558,"proc.ppid":947557,"proc.pvpid":850,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":851}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"sh -c bash -c 'exec /proc/848/fd/3 -l /proc/848/fd/'","proc.exepath":"/bin/sh","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947559,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":852}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"bash -c exec /proc/848/fd/3 -l /proc/848/fd/","proc.exepath":"/usr/bin/bash","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947560,"proc.ppid":947559,"proc.pvpid":852,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":853}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"3 -l /proc/848/fd/","proc.exepath":"/proc/848/fd/3","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947560,"proc.ppid":947559,"proc.pvpid":852,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":853}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"sh -c whoami","proc.exepath":"/bin/sh","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947561,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":854}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"whoami","proc.exepath":"/usr/bin/whoami","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947562,"proc.ppid":947561,"proc.pvpid":854,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":855}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"sh -c bash -c 'echo 123456789 | base64'","proc.exepath":"/bin/sh","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947563,"proc.ppid":947555,"proc.pvpid":848,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":856}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"bash -c echo 123456789 | base64","proc.exepath":"/usr/bin/bash","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947564,"proc.ppid":947563,"proc.pvpid":856,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":857}
{"container.id":"05df91a3e423","evt.type":"execve","proc.cmdline":"base64","proc.exepath":"/usr/bin/base64","proc.is_sid_leader":false,"proc.is_vpgid_leader":false,"proc.pid":947566,"proc.ppid":947564,"proc.pvpid":857,"proc.sid":901354,"proc.sid.exepath":"/usr/bin/bash","proc.sname":"bash","proc.tty":34816,"proc.vpgid":848,"proc.vpgid.exe":"python3","proc.vpgid.exepath":"/usr/bin/python3","proc.vpgid.name":"python3","proc.vpid":859}

```

File: userspace/libsinsp/filterchecks.cpp [hunk 3]
```diff
@@ -2782,6 +2787,174 @@ uint8_t* sinsp_filter_check_thread::extract(sinsp_evt *evt, OUT uint32_t* len, b
 				RETURN_EXTRACT_STRING(m_tstr);
 			}
 		}
+	case TYPE_VPGID_NAME:
+		{
+			bool is_host = true;
+			sinsp_threadinfo* vpgidinfo = NULL;
+			auto flags = tinfo->m_flags;
+			if((flags && flags & PPM_CL_CHILD_IN_PIDNS) || tinfo->m_tid != tinfo->m_vtid)
```
Comment (incertum) [inline userspace/libsinsp/filterchecks.cpp:2795]: @gnosek I did not forget about this detail ...

Comment (Andreagit97) [inline userspace/libsinsp/filterchecks.cpp:2795]: here I would simply leave
```suggestion
			if((tinfo->m_flags & PPM_CL_CHILD_IN_PIDNS) || tinfo->m_tid != tinfo->m_vtid)
			{
			...
			}
```
and maybe we could create a useful `sinsp_threadinfo` method like `is_in_container()` that does exactly this check and return a bool. WDYT?

Comment (Andreagit97) [reply userspace/libsinsp/filterchecks.cpp:2795]: With this helper we could obtain a slightly simpler code

```cpp
case TYPE_VPGID_NAME:
{
	if(!tinfo->is_in_container())
	{
		// Relying on the convention that a process group id is the process id of the process group leader.
		// `threadinfo` lookup only applies when the process is running on the host and not in a pid
		// namespace. However, if the process is running in a pid namespace, we instead traverse the process
		// lineage until we find a match.
		auto vpgidinfo = m_inspector->get_thread_ref(tinfo->m_vpgid, false).get();
		if(vpgidinfo != NULL)
		{
			m_tstr = vpgidinfo->get_comm();
			RETURN_EXTRACT_STRING(m_tstr);
		}
	}

	// This can occur when the process group leader process has exited or if the process
	// is running in a pid namespace and we only have the virtual process group id, as
	// seen from its pid namespace.
	// Find the highest ancestor process that has the same process group id and
	// declare it to be the process group leader.
	sinsp_threadinfo* group_leader = nullptr;

	int64_t vpgid = tinfo->m_vpgid;
	sinsp_threadinfo::visitor_func_t visitor = [vpgid, &group_leader](sinsp_threadinfo* pt)
	{
		if(pt->m_vpgid != vpgid)
		{
			return false;
		}
		group_leader = pt;
		return true;
	};

	tinfo->traverse_parent_state(visitor);

	// mt has been updated to the highest process that has the same process group id.
	// mt's comm is considered the process group leader.
	m_tstr = group_leader->get_comm();
	RETURN_EXTRACT_STRING(m_tstr);
}
```

Not sure it will compile, just to give you an idea of what I have in mind, moreover why do we need to get the main thread? in the end, the `vpgid` should be the same for all threads of the group

**Review by Andreagit97 — COMMENTED** [review]
I think that the new approach works, right now we cannot do something more unfortunately :/ 
if we are not in a container we try to use the table otherwise we fallback in any case to the traverse, just one question about the `main_thread` necessity, see the comments :)

File: userspace/libsinsp/filterchecks.cpp (hunk 3 above)
Comment (incertum) [reply userspace/libsinsp/filterchecks.cpp:2795]: Addressed (slightly) modified to still fall back to parent lineage traversal for host processes where leader may have exited and here we talk only about `is_in_pid_namespace` not what sinsp thinks is a container based on container engine support.

File: userspace/libsinsp/threadinfo.h [hunk 4]
```diff
@@ -113,6 +113,14 @@ class SINSP_PUBLIC sinsp_threadinfo: public libsinsp::state::table_entry
 		return (m_tid == m_pid) || m_flags & PPM_CL_IS_MAIN_THREAD;
 	}
 
+	/*!
+	  \brief Return true if this thread belongs to a pid namespace.
+	*/
+	inline bool is_in_pid_namespace() const
+	{
+		return (m_tid == m_vtid) || m_flags & PPM_CL_CHILD_IN_PIDNS;
```
Comment (Andreagit97) [inline userspace/libsinsp/threadinfo.h:121]:
```suggestion
		return (m_tid != m_vtid) || m_flags & PPM_CL_CHILD_IN_PIDNS;
```

File: userspace/libsinsp/filterchecks.cpp [hunk 5]
```diff
@@ -3537,6 +3680,12 @@ uint8_t* sinsp_filter_check_thread::extract(sinsp_evt *evt, OUT uint32_t* len, b
 	case TYPE_IS_EXE_FROM_MEMFD:
 		m_tbool = tinfo->m_exe_from_memfd;
 		RETURN_EXTRACT_VAR(m_tbool);
+	case TYPE_IS_SID_LEADER:
+		m_tbool = tinfo->m_sid == tinfo->m_vpid ? true : false;
```
Comment (Andreagit97) [inline userspace/libsinsp/filterchecks.cpp:3684]: why do we use 
```bash
tinfo->m_sid == tinfo->m_vpid
```

instead of

```bash
tinfo->m_sid == tinfo->m_pid
```

`m_sid` is referred to the init namespace so probably we should use `m_pid` or am I missing something?

File: userspace/libsinsp/filterchecks.cpp (hunk 3 above)
Comment (Andreagit97) [reply userspace/libsinsp/filterchecks.cpp]: uhm to be honest also the proposed solution :point_up: offers a fallback for the host case :thinking: I won't fight for this but if we can remove nested code (in this case an `if/else`) I would go for it

File: userspace/libsinsp/filterchecks.cpp (hunk 5 above)
Comment (Andreagit97) [reply userspace/libsinsp/filterchecks.cpp:3684]: actually here is correct to use `vpid` since `m_sid` could be also referred to a pid namespace :/

File: userspace/libsinsp/threadinfo.h (hunk 4 above)
Comment (incertum) [reply userspace/libsinsp/threadinfo.h]: oh now that copy and paste error was just fitting :rofl:

**Comment by incertum** [comment]
@FedeDP self-reviewed again and pushed some super minor simplifications and some more text to the field descriptions, also tested it again, see previous example outputs https://github.com/falcosecurity/libs/pull/1178/files#r1247265487. From my perspective it seems to work and should be good to go. Thanks for taking a close look!

File: userspace/libsinsp/filterchecks.cpp [hunk 6]
```diff
@@ -3537,6 +3661,12 @@ uint8_t* sinsp_filter_check_thread::extract(sinsp_evt *evt, OUT uint32_t* len, b
...
```
Comment (FedeDP) [inline userspace/libsinsp/filterchecks.cpp:3665]: Super nit, feel free to skip this :D 
```suggestion
		m_tbool = tinfo->m_sid == tinfo->m_vpid;
```

Same for `TYPE_IS_VPGID_LEADER` below!

**Review by FedeDP — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Review by jasondellaluce — APPROVED** [review]
/approve
